rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,429 @@
+
+Since PIN and touch policies are configured locally on a PIV private key, we cannot delegate the task of setting policies for a private key to the server. Additionally, since these policies depend on Cluster Auth Preference and Role settings, there is no way for us to give the required settings to the user before they login.
+
+To work around this limitiation, we will add another gRPC endpoint `GetPIVAssertionRequirements` for a user to fetch their PIV requirements:
+
+```
+rpc GetPIVAssertionRequirements() returns (PIVAssertionRequirements);
+
+message PIVAssertionRequirements {
+    string Mode = 1;
+    string PINPolicy = 2;
+    string TouchPolicy = 3;
+}
+```
+
+This can be performed after the initial login/signing flow, and before attestation. If the user did not use `--piv` or did not provide strict enough `--piv-pin-policy` or `-piv-touch-policy`, then the user will see a warning message, and `tsh login` will automatically try to generate a valid PIV key and re-perform the login/singing flow with the new key.
+
+### PIV secret management - Management Key, PUX, and PIN
+
+Some PIV operations require [adminstrative access](https://developers.yubico.com/PIV/Introduction/Admin_access.html), which require one or more of the following secrets:
+
+| Name           | size     | default value                                      | function                                  |
+|----------------|----------|----------------------------------------------------|-------------------------------------------|
+| Management Key | 24 bytes | `010203040506070801020304050607080102030405060708` | private key and certificate management    |
+| PIN            | 8 chars  | `123456`                                           | sign and decrypt data, reset pin          |
+| PUK            | 8 chars  | `12345678`                                         | reset PIN when blocked by failed attempts |
+
+The Management Key must be provided during login, and the PIN must be provided during login and during handshakes. Using the default secrets is not recommended, so we will prompt the user for these values when they are needed.
+
+```
+> tsh login --piv
+Enter the PIV card's PIN:
+Enter the PIV card's Management Key:
+```
+
+#### tsh piv configure
+
+We will provide a new command - `tsh piv configure` - so that users can configure their PIV management secrets before they login. This command will prompt the user for the current PIN, PUK, and Management Key, with defaults mentioned. The PIN and PUK will be set from user input, while the Management Key will be randomly generated and stored in [PIN protected metadata](https://pkg.go.dev/github.com/go-piv/piv-go@v1.9.0/piv#YubiKey.Metadata). This way, the user will only need to provide the PIN during login. If the user does not have a management key set in metadata, because they didn't run `tsh piv configure` previously, then they will still be prompted as shown above.
```

Comment:
It's recommended by `piv-go`, or at least this is the intended purpose of [piv.Metadata](https://pkg.go.dev/github.com/go-piv/piv-go@v1.9.0/piv#Metadata).

I think that for the Teleport use case in particular, the management key does not provide any real additional security. The PIN is what ultimately controls access to the PIV card, for signing, decrypting, etc.

The management key essentially only grants access to generating/importing key pairs and importing certificates ([source](https://developers.yubico.com/PIV/Introduction/Admin_access.html)). Since `tsh` is managing keys and certificates, `tsh` will also be verifying that these values haven't been altered. If they are altered, this would just nullify the Teleport certificates which rely on the original private key anyway. 

Furthermore, unless the user is using the default mgm key, they're going to have to store it somewhere, since memorizing a 24 byte key is not feasible. Storing it in the metadata is much better UX and potentially more secure than having the user save it somewhere themselves, like on a sticky note. 

Maybe we should make this feature optional. By default, we will just print out the random management key for the user to store somewhere. If they want to store the mgm key in metadata, we will prompt the user with a warning about how this could affect security if used with other PIV applications, since the mgm key could be easier to obtain.

```
> tsh piv configure --set-mgm-key
Please enter the current management key [blank to use default key]:
New management Key generated: <random-24-byte-key>
Would you like to store this key in PIN protected PIV metadata, so that tsh can find it automatically? (y/N):
// > y
WARNING: This is not recommended if you are using your PIV card with any other applications. Continue? (y/N):
// > y
```

We can also remove `tsh piv configure --get-mgm-key` so that users cannot easily retrieve the mgm key and use it with other applications.

Let me know what you think or if you'd like me to add this reasoning to the RFD.

Reply:
Let's add a nod to the piv.Metadata link you provided, it's valuable information. Up to you if you want to add some of the background in the comment above too.

Removing `tsh piv configure --get-mgm-key` seems reasonable, as long as we let the user know that they _really_ need to jot it down during creation.

About the warning in `tsh piv configure --set-mgm-key`, I think we should word it according to how we perceive the feature. If it's a net positive in UX/security, we go less stern, otherwise a bit more.